lib/wibox/widget/calendar.lua

```diff
@@ -139,8 +139,23 @@ end
 -- @tparam number|nil date.day Date day
 -- @treturn widget Grid layout
 local function create_month(props, date)
-    local num_rows    = 8
-    local num_columns = props.week_numbers and 8 or 7
+    local start_row    = 3
+    local week_start   = props.start_sunday and 1 or 2
+    local last_day     = os.date("*t", os.time{year=date.year, month=date.month+1, day=0})
+    local month_days   = last_day.day
+    local column_fday  = (last_day.wday - month_days + 1 - week_start ) % 7
+
+    local num_columns  = props.week_numbers and 8 or 7
+    local start_column = num_columns - 6
+    local num_rows     = last_day.day // 7
+    if last_day.day > 28 then
+        num_rows = num_rows + 1
+        if column_fday >= 6 then
+            num_rows = num_rows + 1
+        end
+    elseif column_fday > 1 then
+        num_rows = num_rows + 1
+    end
```

Comment:
The math is neat, but requires quite a few magic numbers and a while to get into if you rarely work with calendar calculations.

A short comment explaining the algorithm would make this a lot nicer to skim.

Reply:
True, I'm adding comments.

Reply:
Turns out the math wasn't that neat ! Hopefully that's easier to understand what's happening now.